Consolidate TFM signing

oyvindronningstad opened the pull request:
Use the same signing code for every board.

Depends on #31913

oyvindronningstad committed:
boards: Consolidate all TFM signing code

Place the new signing code in the TFM module CMakeLists.txt.
Make some small tweaks and add a sentence to the docs.

In the process, make a few changes to the signing code:
- Change some names of files created.
- Minimize the number of files created.
- Use hex files instead of bin files. This is so we don't need to know
  the offset when creating hex files from bin files.

Also add signing for MCUBOOT_IMAGE_NUMBER=1 based on the code from the
v2m_musca_b1 board, though, this board does not build with =1 now
because of (I assume) some flash aliasing which places the S and NS
images 0x10000000 apart, where the manual algorithm places them next to
each other. It builds with =2, though.

Signed-off-by: Øyvind Rønningstad <oyvind.ronningstad@nordicsemi.no>

hakonfam commented on boards/CMakeLists.txt:
```diff
@@ -12,3 +12,125 @@ if(EXISTS ${BOARD_DIR}/CMakeLists.txt)
+        --pad-header
+        ${ADD_NS_IMAGE_MIN_VER}
+        -s auto
+        -H 0x400
```
Size of header? Should we have a comment for these arguments?

oyvindronningstad replied:
Used Kconfig var instead.

hakonfam reviewed (approved):
Nice!

ioannisg reviewed (commented):
Is there a way for the common cmake stuff for signing to reside in the TF-M CMakeLists.txt (currently in the module directory, but will move into the main tree). In this way we don't pollute the main boards' CMakeLists.txt file.

That's a question for @tejlmand

ioannisg commented on modules/Kconfig.tfm:
```diff
@@ -24,6 +24,7 @@ menuconfig BUILD_WITH_TFM
 	depends on TRUSTED_EXECUTION_NONSECURE
 	depends on TFM_BOARD != ""
 	depends on ARM_TRUSTZONE_M
+	imply BUILD_OUTPUT_HEX # For signing.
```
No need for the inline comment, IMHO

oyvindronningstad replied:
Done

tejlmand commented:
> Is there a way for the common cmake stuff for signing to reside in the TF-M CMakeLists.txt (currently in the module directory, but will move into the main tree). In this way we don't pollute the main boards' CMakeLists.txt file.

I had a quick look, and initial feeling is that this should indeed be doable.
Haven't noticed anything that makes me think it should not be possible, but of course we need to ensure the post build commands are still executed in correct order, for those that are depending on other part.

As side note, I've noticed increased use of the extra post build command property and to ensure correct dependency handling, some of those should probably be changed into custom commands for proper dependency chain.

tejlmand commented on modules/trusted-firmware-m/Kconfig:
```diff
@@ -25,6 +25,7 @@ menuconfig BUILD_WITH_TFM
 	depends on TRUSTED_EXECUTION_NONSECURE
 	depends on TFM_BOARD != ""
 	depends on ARM_TRUSTZONE_M
+	imply BUILD_OUTPUT_HEX
```
is `imply` strong enough here ?
If there is not hex output, will the following build / signing / merge stages work as expected ?

oyvindronningstad replied:
No, it will fail with "file not found" or something.

oyvindronningstad replied:
Is `select` better?

tejlmand replied:
in that case, yes.

oyvindronningstad replied:
done

ioannisg reviewed (changes requested):
Looks ok to me, but we probably need to fix the QEMU running for the integration samples, which seems to me it's currently broken.
I think something might be wrong with the executable paths @oyvindronningstad

ioannisg reviewed (commented):
I think this needs some extra work @oyvindronningstad , particularly, for the QEMU target, which seems to be using the tfm_qemu.bin  and .hex respectively.

This is something we can clean up in the samples, but we also need to fix the consolidation, to work for the mps2 target, because even if i switch samples to use tfm_merged.hex, the qemu does not work.

oyvindronningstad commented:
Ok, so just copying tfm_merged.hex to qemu.hex will fix it? We can do that of course. Is there a way to check whether qemu is available and only create it for those targets?

ioannisg committed:
samples: tfm integration: rename qemu targets after refactor

Now that we have consolidated tf-m signing, the combined
binary is always named tfm_merged, so we need to modify
the QEMU targets in the tfm-integration samples.

Signed-off-by: Ioannis Glaropoulos <Ioannis.Glaropoulos@nordicsemi.no>

ioannisg commented:
@oyvindronningstad i pushed a commit that does it (hopefully). Had not seen your comment , sorry.

oyvindronningstad commented:
@ioannisg That looks even better, thanks!
